sirepo/job_driver/__init__.py

```diff
@@ -88,54 +85,12 @@ def __init__(self, req):
         self.__instances[self._agentId] = self
         pkdlog('{}', self)
 
-    def cpu_slot_free(self):
-        if not self.cpu_slot:
-            return
-        self.cpu_slot_q.task_done()
-        self.cpu_slot_q.put_nowait(self.cpu_slot)
-        self.cpu_slot = None
-        self._cpu_slot_alloc_time = None
-
-    def cpu_slot_free_one(self):
-        if self.cpu_slot_q.qsize() > 0:
-            # available slots, don't need to free
-            return
-        # This is not fair scheduling, but good enough for now.
-        # least recently used and not in use
-        d = sorted(
-            filter(
-                lambda x: bool(x.cpu_slot and not x.ops),
-                self.cpu_slot_peers(),
-            ),
-            key=lambda x: x._cpu_slot_alloc_time,
-        )
-        if d:
-            d[0].cpu_slot_free()
-
-    async def cpu_slot_ready(self):
-        if self.cpu_slot:
-            return
-        try:
-            self.cpu_slot = self.cpu_slot_q.get_nowait()
-        except tornado.queues.QueueEmpty:
-            self.cpu_slot_free_one()
-            pkdlog('{} await cpu_slot_q.get()', self)
-            self.cpu_slot = await self.cpu_slot_q.get()
-            raise job_supervisor.Awaited()
-        finally:
-            self._cpu_slot_alloc_time = time.time()
-
     def destroy_op(self, op):
         """Clear our op and (possibly) free cpu slot"""
         self.ops.pkdel(op.opId)
-        if not self.ops:
-            # might free our cpu slot if no other ops
-            self.cpu_slot_free_one()
+        op.cpu_slot.free()
```

Comment:
This should be guarded with an if too. It may be better to move the assignment of the slot proxies to the op in op initialization so we can remove the guard if's.

Reply:
There's always a cpu_slot. I changed sbatch to create cpu_slots (for every driver) the same number of ops. This will work properly because ops is the restrictive case.